File: trueledge/src/lib/vouchers/voucher-types.ts
import type { VoucherType } from "@/types/database.types";

/**
 * Voucher type registry — the single source of truth for the Tally-style
 * function-key shortcuts and the posting shape of each voucher type.
 *
 * `mode` drives both the entry UI and the double-entry generation in
 * `src/lib/actions/voucher.ts`:
 *
 *   - "settlement"  Two-sided cash movement (Contra, Payment, Receipt). One
 *                   ledger is paid from, the other paid to. No tax, no items.
 *   - "journal"     Free-form Dr/Cr grid. The accountant states both sides, so
 *                   every line carries an explicit debit or credit.
 *   - "invoice"     Commercial document (Sales, Purchase). Line items with
 *                   quantity/rate/tax; the party control account is the
 *                   balancing side and VAT posts to the tax code's account.
 */
export type VoucherMode = "settlement" | "journal" | "invoice";

/** Direction of the trade for invoice-mode vouchers. */
export type TradeDirection = "outward" | "inward";

export interface VoucherTypeConfig {
  type: VoucherType;
  /** Tally function key, e.g. "F5". */
  shortcut: string;
  label: string;
  /** Short hint shown under the tab. */
  hint: string;
  mode: VoucherMode;
  /**
   * For settlement vouchers: which side the "from" ledger sits on.
   * Payment credits the bank (money leaves); Receipt debits it (money arrives).
   */
  direction?: TradeDirection;
  /** Whether the voucher requires a party (customer/supplier). */
  requiresParty: boolean;
  /** Whether line items carry tax. */
  hasTax: boolean;
}

export const VOUCHER_TYPES: readonly VoucherTypeConfig[] = [
  {
    type: "contra",
    shortcut: "F4",
    label: "Contra",
    hint: "Bank ↔ Cash transfer",
    mode: "settlement",
    requiresParty: false,
    hasTax: false,
  },
  {
    type: "payment_voucher",
    shortcut: "F5",
    label: "Payment",
    hint: "Pay supplier or expense",
    mode: "settlement",
    direction: "outward",
    requiresParty: false,
    hasTax: false,
  },
  {
    type: "receipt_voucher",
    shortcut: "F6",
    label: "Receipt",
    hint: "Receive from customer",
    mode: "settlement",
    direction: "inward",
    requiresParty: false,
    hasTax: false,
  },
  {
    type: "journal_voucher",
    shortcut: "F7",
    label: "Journal",
    hint: "Manual Dr/Cr adjustment",
    mode: "journal",
    requiresParty: false,
    hasTax: false,
  },
  {
    type: "sales_invoice",
    shortcut: "F8",
    label: "Sales",
    hint: "Tax invoice to customer",
    mode: "invoice",
    direction: "outward",
    requiresParty: true,
    hasTax: true,
  },
  {
    type: "purchase_bill",
    shortcut: "F9",
    label: "Purchase",
    hint: "Supplier bill",
    mode: "invoice",
    direction: "inward",
    requiresParty: true,
    hasTax: true,
  },
] as const;

const BY_TYPE = new Map(VOUCHER_TYPES.map((v) => [v.type, v]));
const BY_SHORTCUT = new Map(VOUCHER_TYPES.map((v) => [v.shortcut, v]));

export function getVoucherConfig(type: VoucherType): VoucherTypeConfig | undefined {
  return BY_TYPE.get(type);
}

export function getVoucherConfigByShortcut(key: string): VoucherTypeConfig | undefined {
  return BY_SHORTCUT.get(key.toUpperCase());
}

/** Voucher types reachable from the unified entry screen. */
export function isEntryVoucherType(type: string): type is VoucherType {
  return BY_TYPE.has(type as VoucherType);
}

File: xorvaerp/backend/Modules/Xorva.Modules.Accounting/Vouchers/Common/VoucherTypes.cs
using Xorva.Modules.Accounting.Enums;

namespace Xorva.Modules.Accounting.Vouchers.Common;

/// <summary>How a voucher family is entered and turned into ledger lines (ported from TrueLedge voucher-types.ts).</summary>
public enum VoucherMode
{
    /// <summary>Contra / Payment / Receipt — two-ledger money movement, no tax.</summary>
    Settlement,
    /// <summary>Journal — free Dr/Cr grid.</summary>
    Journal,
    /// <summary>Sales / Purchase — item lines with VAT and a balancing party (control-account) line.</summary>
    Invoice
}

public enum TradeDirection { Inward, Outward }

/// <summary>Static description of one Tally voucher family — function key, mode, party rules, prefix.</summary>
public sealed record VoucherTypeConfig(
    VoucherType Type,
    string Shortcut,
    string Label,
    string Description,
    VoucherMode Mode,
    bool RequiresParty,
    TradeDirection? Direction,
    string Prefix);

/// <summary>The F4–F9 catalogue. Prefixes MUST agree with <c>accounting.voucher_prefix</c> (Sql/Accounting/0003).</summary>
public static class VoucherTypes
{
    public static readonly IReadOnlyList<VoucherTypeConfig> All =
    [
        new(VoucherType.Contra,       "F4", "Contra",   "Move money between bank and cash ledgers.",                    VoucherMode.Settlement, false, null,                    "CT"),
        new(VoucherType.Payment,      "F5", "Payment",  "Money going out — supplier bills, expenses, salaries.",         VoucherMode.Settlement, false, TradeDirection.Outward,  "PV"),
        new(VoucherType.Receipt,      "F6", "Receipt",  "Money coming in — customer collections, refunds, deposits.",    VoucherMode.Settlement, false, TradeDirection.Inward,   "RV"),
        new(VoucherType.Journal,      "F7", "Journal",  "Adjustments, accruals, provisions — any Dr/Cr entry.",          VoucherMode.Journal,    false, null,                    "JV"),
        new(VoucherType.SalesInvoice, "F8", "Sales",    "Tax invoice to a customer with VAT lines.",                     VoucherMode.Invoice,    true,  TradeDirection.Outward,  "SI"),
        new(VoucherType.PurchaseBill, "F9", "Purchase", "Supplier tax invoice with recoverable input VAT.",              VoucherMode.Invoice,    true,  TradeDirection.Inward,   "PB"),
        // Not entered through the F-key screen, but they are valid voucher rows (reversals, migrated notes, opening balances).
        new(VoucherType.CreditNote,   "",   "Credit Note",     "Credit issued against a sales invoice.",                 VoucherMode.Invoice,    true,  TradeDirection.Inward,   "CN"),
        new(VoucherType.DebitNote,    "",   "Debit Note",      "Debit raised against a supplier bill.",                  VoucherMode.Invoice,    true,  TradeDirection.Outward,  "DN"),
        new(VoucherType.OpeningBalance, "", "Opening Balance", "Books-beginning balances.",                             VoucherMode.Journal,    false, null,                    "OB"),
    ];

    private static readonly Dictionary<VoucherType, VoucherTypeConfig> ByType = All.ToDictionary(v => v.Type);

    public static VoucherTypeConfig For(VoucherType type) => ByType[type];

    /// <summary>The six families the entry screen offers (F4–F9), in key order.</summary>
    public static IEnumerable<VoucherTypeConfig> EntryTypes => All.Where(v => v.Shortcut.Length > 0);
}
